k8s.io/apimachinery/pkg/util/wait.PollUntilContextCancel immediately executes condition twice

### What happened?

While trying to update https://github.com/flatcar/flatcar-linux-update-operator/blob/7a1b0ff2769af91e8ff2bae82fc93f2f5baf0b0c/pkg/agent/agent.go#L328 to use latest version of `wait` package and to migrate away from deprecated `PollImmediateUntil`, function, our tests fail as it seems new `PollUntilContextCancel` function immediately executes given condition twice, which is not mentioned anywhere.

Result of executed test specified below:
```console
$ go test . -v
=== RUN   Test_test
=== PAUSE Test_test
=== CONT  Test_test
=== RUN   Test_test/PollUntilContextCancel
    prog_test.go:25: Executed at 2009-11-10 23:00:00 +0000 UTC m=+0.000000001
    prog_test.go:25: Executed at 2009-11-10 23:00:00 +0000 UTC m=+0.000000001
    prog_test.go:25: Unexpected early timestamp
=== RUN   Test_test/PollImmediateUntil
    prog_test.go:31: Executed at 2009-11-10 23:00:00 +0000 UTC m=+0.000000001
    prog_test.go:31: Executed at 2009-11-10 23:00:02 +0000 UTC m=+2.000000001
--- FAIL: Test_test (3.00s)
    --- FAIL: Test_test/PollUntilContextCancel (0.00s)
    --- PASS: Test_test/PollImmediateUntil (3.00s)
FAIL
```

### What did you expect to happen?

Since `k8s.io/apimachinery/pkg/util/wait.PollImmediateUntil` has been deprecated and it recommends using `k8s.io/apimachinery/pkg/util/wait.PollUntilContextCancel`, I expected it to have the equivalent semantics to allow easy migration from deprecated functions, as no changelog or documentation mentions new semantics.

### How can we reproduce it (as minimally and precisely as possible)?

Playground: https://go.dev/play/p/qQBnWM4nufF

Local:
1. `go mod init foo && go get k8s.io/apimachinery/pkg/util/wait`
2. Create `main_test.go` file with the following content:
```go
package main

import (
	"context"
	"testing"
	"time"

	"k8s.io/apimachinery/pkg/util/wait"
)

const (
	testTimeout           = 3 * time.Second
	callsExpectedInterval = 2 * time.Second
)

func Test_test(t *testing.T) {
	t.Parallel()

	ctx, cancel := context.WithTimeout(context.Background(), testTimeout)
	t.Cleanup(cancel)

	t.Run("PollUntilContextCancel", func(t *testing.T) {
		f := testIntervals(t)

		wait.PollUntilContextCancel(ctx, callsExpectedInterval, true, func(context.Context) (bool, error) { f(); return false, nil })
	})

	t.Run("PollImmediateUntil", func(t *testing.T) {
		f := testIntervals(t)

		wait.PollImmediateUntil(callsExpectedInterval, func() (bool, error) { f(); return false, nil }, ctx.Done())
	})
}

func testIntervals(t *testing.T) func() {
	t.Helper()

	timestamps := make(chan time.Time, 1)

	timestamps <- time.Now().Add(-1 * 2 * callsExpectedInterval)

	return func() {
		t.Helper()

		t.Logf("Executed at %s", time.Now())

		if timestamp := <-timestamps; timestamp.After(time.Now().Add(-1 * callsExpectedInterval / 2)) {
			t.Fatalf("Unexpected early timestamp")
		}

		timestamps <- time.Now()
	}
}
```
3. Run as `go test . -v`.

### Anything else we need to know?

_No response_

### Kubernetes version

<details>

```console
$ kubectl version
# paste output here
```

</details>


### Cloud provider

<details>

</details>


### OS version

<details>

```console
# On Linux:
$ cat /etc/os-release
# paste output here
$ uname -a
# paste output here

# On Windows:
C:\> wmic os get Caption, Version, BuildNumber, OSArchitecture
# paste output here
```

</details>


### Install tools

<details>

</details>


### Container runtime (CRI) and version (if applicable)

<details>

</details>


### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

--- a/staging/src/k8s.io/apimachinery/pkg/util/wait/loop.go
+++ b/staging/src/k8s.io/apimachinery/pkg/util/wait/loop.go
@@ -40,6 +40,10 @@ func loopConditionUntilContext(ctx context.Context, t Timer, immediate, sliding 
 	var timeCh <-chan time.Time
 	doneCh := ctx.Done()
 
+	if !sliding {
+		timeCh = t.C()
+	}
+
 	// if immediate is true the condition is
 	// guaranteed to be executed at least once,
 	// if we haven't requested immediate execution, delay once
@@ -50,17 +54,27 @@ func loopConditionUntilContext(ctx context.Context, t Timer, immediate, sliding 
 		}(); err != nil || ok {
 			return err
 		}
-	} else {
+	}
+
+	if sliding {
 		timeCh = t.C()
+	}
+
+	for {
+
+		// Wait for either the context to be cancelled or the next invocation be called
 		select {
 		case <-doneCh:
 			return ctx.Err()
 		case <-timeCh:
 		}
-	}
 
-	for {
-		// checking ctx.Err() is slightly faster than checking a select
+		// IMPORTANT: Because there is no channel priority selection in golang
+		// it is possible for very short timers to "win" the race in the previous select
+		// repeatedly even when the context has been canceled.  We therefore must
+		// explicitly check for context cancellation on every loop and exit if true to
+		// guarantee that we don't invoke condition more than once after context has
+		// been cancelled.
 		if err := ctx.Err(); err != nil {
 			return err
 		}
@@ -76,22 +90,6 @@ func loopConditionUntilContext(ctx context.Context, t Timer, immediate, sliding 
 		}
 		if sliding {
 			t.Next()
-		}
-
-		if timeCh == nil {
-			timeCh = t.C()
-		}
-
-		// NOTE: b/c there is no priority selection in golang
-		// it is possible for this to race, meaning we could
-		// trigger t.C and doneCh, and t.C select falls through.
-		// In order to mitigate we re-check doneCh at the beginning
-		// of every loop to guarantee at-most one extra execution
-		// of condition.
-		select {
-		case <-doneCh:
-			return ctx.Err()
-		case <-timeCh:
 		}
 	}
 }

--- a/staging/src/k8s.io/apimachinery/pkg/util/wait/loop_test.go
+++ b/staging/src/k8s.io/apimachinery/pkg/util/wait/loop_test.go
@@ -206,42 +206,34 @@ func Test_loopConditionUntilContext_semantic(t *testing.T) {
 	}
 
 	for _, test := range tests {
-		for _, immediate := range []bool{true, false} {
-			t.Run(fmt.Sprintf("immediate=%t", immediate), func(t *testing.T) {
-				for _, sliding := range []bool{true, false} {
-					t.Run(fmt.Sprintf("sliding=%t", sliding), func(t *testing.T) {
-						t.Run(test.name, func(t *testing.T) {
-							contextFn := test.context
-							if contextFn == nil {
-								contextFn = defaultContext
-							}
-							ctx, cancel := contextFn()
-							defer cancel()
+		t.Run(test.name, func(t *testing.T) {
+			contextFn := test.context
+			if contextFn == nil {
+				contextFn = defaultContext
+			}
+			ctx, cancel := contextFn()
+			defer cancel()
 
-							timer := Backoff{Duration: time.Microsecond}.Timer()
-							attempts := 0
-							err := loopConditionUntilContext(ctx, timer, test.immediate, test.sliding, func(_ context.Context) (bool, error) {
-								attempts++
-								defer func() {
-									if test.cancelContextAfter > 0 && test.cancelContextAfter == attempts {
-										cancel()
-									}
-								}()
-								return test.callback(attempts)
-							})
-
-							if test.errExpected != err {
-								t.Errorf("expected error: %v but got: %v", test.errExpected, err)
-							}
-
-							if test.attemptsExpected != attempts {
-								t.Errorf("expected attempts count: %d but got: %d", test.attemptsExpected, attempts)
-							}
-						})
-					})
-				}
+			timer := Backoff{Duration: time.Microsecond}.Timer()
+			attempts := 0
+			err := loopConditionUntilContext(ctx, timer, test.immediate, test.sliding, func(_ context.Context) (bool, error) {
+				attempts++
+				defer func() {
+					if test.cancelContextAfter > 0 && test.cancelContextAfter == attempts {
+						cancel()
+					}
+				}()
+				return test.callback(attempts)
 			})
-		}
+
+			if test.errExpected != err {
+				t.Errorf("expected error: %v but got: %v", test.errExpected, err)
+			}
+
+			if test.attemptsExpected != attempts {
+				t.Errorf("expected attempts count: %d but got: %d", test.attemptsExpected, attempts)
+			}
+		})
 	}
 }
 

--- a/staging/src/k8s.io/apimachinery/pkg/util/wait/loop_test.go
+++ b/staging/src/k8s.io/apimachinery/pkg/util/wait/loop_test.go
@@ -99,6 +99,7 @@ func Test_loopConditionUntilContext_semantic(t *testing.T) {
 		cancelContextAfter int
 		attemptsExpected   int
 		errExpected        error
+		timer              Timer
 	}{
 		{
 			name: "condition successful is only one attempt",
@@ -203,6 +204,54 @@ func Test_loopConditionUntilContext_semantic(t *testing.T) {
 			attemptsExpected: 0,
 			errExpected:      context.DeadlineExceeded,
 		},
+		{
+			name:      "context canceled before the second execution and immediate",
+			immediate: true,
+			context: func() (context.Context, context.CancelFunc) {
+				return context.WithTimeout(context.Background(), time.Second)
+			},
+			callback: func(attempts int) (bool, error) {
+				return false, nil
+			},
+			attemptsExpected: 1,
+			errExpected:      context.DeadlineExceeded,
+			timer:            Backoff{Duration: 2 * time.Second}.Timer(),
+		},
+		{
+			name:      "immediate and long duration of condition and sliding false",
+			immediate: true,
+			sliding:   false,
+			context: func() (context.Context, context.CancelFunc) {
+				return context.WithTimeout(context.Background(), time.Second)
+			},
+			callback: func(attempts int) (bool, error) {
+				if attempts >= 4 {
+					return true, nil
+				}
+				time.Sleep(time.Second / 5)
+				return false, nil
+			},
+			attemptsExpected: 4,
+			timer:            Backoff{Duration: time.Second / 5, Jitter: 0.001}.Timer(),
+		},
+		{
+			name:      "immediate and long duration of condition and sliding true",
+			immediate: true,
+			sliding:   true,
+			context: func() (context.Context, context.CancelFunc) {
+				return context.WithTimeout(context.Background(), time.Second)
+			},
+			callback: func(attempts int) (bool, error) {
+				if attempts >= 4 {
+					return true, nil
+				}
+				time.Sleep(time.Second / 5)
+				return false, nil
+			},
+			errExpected:      context.DeadlineExceeded,
+			attemptsExpected: 3,
+			timer:            Backoff{Duration: time.Second / 5, Jitter: 0.001}.Timer(),
+		},
 	}
 
 	for _, test := range tests {
@@ -214,7 +263,10 @@ func Test_loopConditionUntilContext_semantic(t *testing.T) {
 			ctx, cancel := contextFn()
 			defer cancel()
 
-			timer := Backoff{Duration: time.Microsecond}.Timer()
+			timer := test.timer
+			if timer == nil {
+				timer = Backoff{Duration: time.Microsecond}.Timer()
+			}
 			attempts := 0
 			err := loopConditionUntilContext(ctx, timer, test.immediate, test.sliding, func(_ context.Context) (bool, error) {
 				attempts++

